### PR TITLE
Optimize slotted runtime result production

### PR DESCRIPTION
**Pull request by henriknyman**
Make ProduceResultSlottedPipe return an iterator that can materialize
the results directly into an array of columns that can be consumed by
Bolt directly without going through a generic Map interface.
This lets us avoid unnecessary copies and hash map creations.

- Introduce an ArrayResultExecutionContext that implements
QueryResult.Record, and stores the columns in an ordered array
for fast direct consumption, with support for the old Map interface
as a secondary concern.

- Use an instance cache for ArrayResultExecutionContext to reduce the
overhead of array creation and initialization.

### CODE REVIEW

File: enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/ExecutionContextInstanceCache.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+package org.neo4j.cypher.internal.compatibility.v3_4.runtime
+
+import org.neo4j.cypher.internal.runtime.interpreted.ExecutionContext
+
+trait ExecutionContextInstanceCache[T, C <: ExecutionContext] {
+  self: T =>
+
+  type ExecutionContextConstructor = T => C
+
+  private var freeExecutionContextInstance: C = null.asInstanceOf[C]
```
Comment (pontusmelke) [inline enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/ExecutionContextInstanceCache.scala]: Having vars in a trait looks dodgy to me, can't solve this nicer?

File: enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/ExecutionContextInstanceCache.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compatibility.v3_4.runtime
+
+import org.neo4j.cypher.internal.runtime.interpreted.ExecutionContext
+
+trait ExecutionContextInstanceCache[T, C <: ExecutionContext] {
```
Comment (pontusmelke) [inline enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/ExecutionContextInstanceCache.scala]: Can we have some docs here explaining the purpose of this thing

**Review by pontusmelke — CHANGES_REQUESTED** [review]
I am not crazy about the solution with having var on a trait, can we have something a little nicer. Some more comments on some things would be nice as well

**Comment by henriknyman** [comment]
@pontusmelke `ExecutionContextInstanceCache` used to be shared code with another usage place with slightly more complicated module dependency in the old PR (and the new execution context pooling branch). I guess I tried to get away with keeping it here but failed ;D
For now with only a single implementation it is bloat, so I am inlining it in the factory instead.

File: enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/ExecutionContextInstanceCache.scala (hunk 1 above)
Comment (henriknyman) [reply enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/ExecutionContextInstanceCache.scala]: This trait is deleted now, but the point was for the using class to not have to be bothered with the implementation. Would you have been more ok to call it an abstract class? Wouldn't that just unnecessarily limit the mix-in possibilities?

**Review by pontusmelke — APPROVED** [review]

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/PipeExecutionResult.scala [hunk 3]
```diff
@@ -109,7 +109,11 @@ class PipeExecutionResult(val result: ResultIterator,
 
   def accept[EX <: Exception](visitor: QueryResultVisitor[EX]): Unit = {
     try {
-      javaValues.feedIteratorToVisitable(result.map(r => fieldNames.map(r))).accept(visitor)
+      val maybeRecordIterator = result.recordIterator
```
Comment (systay) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/PipeExecutionResult.scala:112]: tiny Scala-beautification:

```
result.recordIterator match {
  case Some(recordIterator) => javaValues.feedQueryResultRecordIteratorToVisitable(recordIterator).accept(visitor)
  case None => javaValues.feedIteratorToVisitable(result.map(r => fieldNames.map(r))).accept(visitor)
}
```

**Comment by systay** [comment]
Hehe... Didn't realise it had already been merged :)

**Pull request merged**